diff --git a/src/CodeCoverage.php b/src/CodeCoverage.php
index 2c92ae2..69fbb9d 100644
--- a/src/CodeCoverage.php
+++ b/src/CodeCoverage.php
@@ -194,7 +194,17 @@ public function getData(bool $raw = false): array
             $this->addUncoveredFilesFromWhitelist();
         }
 
-        return $this->data;
+        $data = [];
+        if ($restoreCallback = \PHPUnit\Util\FileLoader::getFilenameRestoreCallback()) {
+            foreach ($this->data as $file => $lines) {
+                $file = $restoreCallback($file);
+                $data[$file] = $lines;
+            }
+        } else {
+            $data = $this->data;
+        }
+
+        return $data;
     }
 
     /**
diff --git a/src/Filter.php b/src/Filter.php
index b3c2d2d..1dc34a0 100644
--- a/src/Filter.php
+++ b/src/Filter.php
@@ -48,6 +48,10 @@ public function addDirectoryToWhitelist(string $directory, string $suffix = '.ph
      */
     public function addFileToWhitelist(string $filename): void
     {
+        if ($rewriteCallback = \PHPUnit\Util\FileLoader::getFilenameRewriteCallback()) {
+            $filename = $rewriteCallback($filename);
+        }
+
         $filename = \realpath($filename);
 
         if (!$filename) {
@@ -87,6 +91,10 @@ public function removeDirectoryFromWhitelist(string $directory, string $suffix =
      */
     public function removeFileFromWhitelist(string $filename): void
     {
+        if ($rewriteCallback = \PHPUnit\Util\FileLoader::getFilenameRewriteCallback()) {
+            $filename = $rewriteCallback($filename);
+        }
+
         $filename = \realpath($filename);
 
         if (!$filename || !isset($this->whitelistedFiles[$filename])) {
