diff --git a/src/Framework/TestSuite.php b/src/Framework/TestSuite.php
index 83c9d9a1c..dc9d900b4 100644
--- a/src/Framework/TestSuite.php
+++ b/src/Framework/TestSuite.php
@@ -376,8 +376,13 @@ public function addTestFile(string $filename): void
                         $e
                     );
                 }
+                $classFile = $class->getFileName();
 
-                if ($class->getFileName() == $filename) {
+                if ($restoreCallback = Fileloader::getFilenameRestoreCallback()) {
+                    $classFile = $restoreCallback($classFile);
+                }
+
+                if ($classFile == $filename) {
                     $newClasses = [$className];
                     unset($this->foundClasses[$i]);
 
diff --git a/src/Runner/StandardTestSuiteLoader.php b/src/Runner/StandardTestSuiteLoader.php
index 47ad010f4..8fcfd0465 100644
--- a/src/Runner/StandardTestSuiteLoader.php
+++ b/src/Runner/StandardTestSuiteLoader.php
@@ -83,6 +83,10 @@ public function load(string $suiteClassName, string $suiteClassFile = ''): Refle
 
                 $classFile = $class->getFileName();
 
+                if ($restoreCallback = Fileloader::getFilenameRestoreCallback()) {
+                    $classFile = $restoreCallback($classFile);
+                }
+
                 if ($class->isSubclassOf($testCaseClass) && !$class->isAbstract()) {
                     $suiteClassName = $loadedClass;
                     $testCaseClass  = $loadedClass;
@@ -125,7 +129,13 @@ public function load(string $suiteClassName, string $suiteClassFile = ''): Refle
                 );
             }
 
-            if ($class->getFileName() == \realpath($suiteClassFile)) {
+            $classFile = $class->getFileName();
+
+            if ($restoreCallback = Fileloader::getFilenameRestoreCallback()) {
+                $classFile = $restoreCallback($classFile);
+            }
+
+            if ($classFile == \realpath($suiteClassFile)) {
                 return $class;
             }
         }
diff --git a/src/Util/FileLoader.php b/src/Util/FileLoader.php
index 2c5f7ca99..9ffc0c08e 100644
--- a/src/Util/FileLoader.php
+++ b/src/Util/FileLoader.php
@@ -16,6 +16,40 @@
  */
 final class FileLoader
 {
+    /** @var Callable */
+    private static $filename_rewrite_callback;
+
+    /** @var Callable */
+    private static $filename_restore_callback;
+
+    /**
+     * Provide callback for rewriting test file names that is called when loading suite files
+     * @param callable $callback (source_filename => rewritten_filename)
+     */
+    public static function setFilenameRewriteCallback(Callable $callback)
+    {
+        self::$filename_rewrite_callback = $callback;
+    }
+
+    public static function getFilenameRewriteCallback()
+    {
+        return self::$filename_rewrite_callback;
+    }
+
+    /**
+     * Provide callback for restoring rewritten test file names
+     * @param callable $callback (rewritten_filename => source_filename)
+     */
+    public static function setFilenameRestoreCallback(Callable $callback)
+    {
+        self::$filename_restore_callback = $callback;
+    }
+
+    public static function getFilenameRestoreCallback()
+    {
+        return self::$filename_restore_callback;
+    }
+
     /**
      * Checks if a PHP sourcecode file is readable. The sourcecode file is loaded through the load() method.
      *
@@ -56,7 +90,12 @@ public static function load(string $filename): void
     {
         $oldVariableNames = \array_keys(\get_defined_vars());
 
-        include_once $filename;
+        if ($cb = self::$filename_rewrite_callback) {
+            $new_filename = $cb($filename);
+            include_once $new_filename;
+        } else {
+            include_once $filename;
+        }
 
         $newVariables     = \get_defined_vars();
 
diff --git a/src/Util/Filter.php b/src/Util/Filter.php
index b37708d87..24b9b4570 100644
--- a/src/Util/Filter.php
+++ b/src/Util/Filter.php
@@ -17,11 +17,26 @@
  */
 final class Filter
 {
+    private static $custom_stacktrace_callback;
+
+    /**
+     * Set custom stack trace callback that can deal with rewritten files and extra stack trace lines
+     * @param callable $callback (Exception $e, $asString = true): string
+     */
+    public static function setCustomStackTraceCallback(Callable $callback)
+    {
+        self::$custom_stacktrace_callback = $callback;
+    }
+
     /**
      * @throws Exception
      */
     public static function getFilteredStacktrace(\Throwable $t): string
     {
+        if ($stackTraceCb = self::$custom_stacktrace_callback) {
+            return $stackTraceCb($t);
+        }
+
         $prefix = false;
         $script = \realpath($GLOBALS['_SERVER']['SCRIPT_NAME']);
 
